framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,228 @@
+
+   Fields:
+
+      Type
+                     1 for Source Link-layer Address
+                     2 for Target Link-layer Address
+      Length         The length of the option (including the type and
+                     length fields) in units of 8 octets.  For example,
+                     the length for IEEE 802 addresses is 1
+                     [IPv6-ETHER].
+
+      Link-Layer Address
+                     The variable length link-layer address.
+
+                     The content and format of this field (including
+                     byte and bit ordering) is expected to be specified
+                     in specific documents that describe how IPv6
+                     operates over different link layers.  For instance,
+                     [IPv6-ETHER].
+
+   Description
+                     The Source Link-Layer Address option contains the
+                     link-layer address of the sender of the packet.  It
+                     is used in the Neighbor Solicitation, Router
+                     Solicitation, and Router Advertisement packets.
+
+                     The Target Link-Layer Address option contains the
+                     link-layer address of the target.  It is used in
+                     Neighbor Advertisement and Redirect packets.
+
+                     These options MUST be silently ignored for other
+                     Neighbor Discovery messages.
+*/
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct LinkLayerAddressOption {
+    option_type: u8,
+    option_length: u8,
+    addr: MacAddress,
+}
+
+impl NdpOption for LinkLayerAddressOption {
+    fn get_type() -> NdpOptionType {
+        NdpOptionType::SourceLinkLayerAddress
+    }
+}
+
+/// Provides functions to implementors that allow the implementor to scan their NDP message for
+/// NDP options.  The type of options available are specific to the type of NDP Message
+/// Ensures that implementors support NdpMessageContents, which further enforces EndOffset + Sized
+/// which we require to do parsing
+pub trait NdpOptions: NdpMessageContents {
+    /// Attempts to lookup the source link layer address from the underlying NDP Message
+    fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let opt = self.option_scan::<LinkLayerAddressOption>(payload_length);
+        opt.map(|sll| &sll.addr)
+    }
+
+    /// Attempts to lookup the mtu option from the underlying NDP Message
+    fn get_mtu_option(&self, payload_length: u16) -> Option<u32> {
+        let mtu_option = self
+            .option_scan::<MtuOption>(payload_length)
+            .map(|mtu| mtu.get_mtu());
+        mtu_option
+    }
+
+    /// Searches through a packet for an option, returns the first it finds or none
+    /// Relies on the caller to specify T, and they must know that the type can be parsed...
+    /// this is unfortunate, but I do not have a compile time way as the NDPOptionType is something
```

Comment:
can remove the comments here about this being unfortunate since we addressed it

Reply:
done